stop_and_wait_plugin/src/stop_and_wait_plugin.cpp

```diff
@@ -43,432 +43,294 @@
+
+  double remaining_distance = stop_location - starting_downtrack; // TODO review. Should the remaining distance be in the middle of the stopping buffer?
+  double target_accel = config_.accel_limit_multiplier * config_.accel_limit;
+  double req_dist =
+      starting_speed / (2.0 * target_accel);  // Distance needed to go from current speed to 0 at target accel
+
+  if (req_dist > remaining_distance)
+  {
+    ROS_DEBUG_STREAM("Target Accel Update From: " << target_accel);
+    target_accel = starting_speed / (2.0 * remaining_distance);  // If we cannot reach the end point it the required
+                                                                 // distance update the accel target
+    ROS_DEBUG_STREAM("Target Accel Update To: " << target_accel);
+  }
+
+  final_points.reserve(points.size());
+
+  PointSpeedPair prev_pair = points.back();
+  prev_pair.speed = 0.0;
+  final_points.push_back(prev_pair);  // Store the points in reverse
+
+  bool reached_end = false;
+  for (int i = points.size() - 2; i >= 0; i--)
+  {  // NOTE: Do not use size_t for i type here as -- with > 0 will result in overflow
+
+    double v_i = prev_pair.speed;
+
+    if (reached_end || v_i >= starting_speed)
+    {  // We are walking backward, so if the prev speed is greater than or equal to the starting speed then we are done
+       // backtracking
+      reached_end = true;
+      PointSpeedPair pair = points[i];
+      pair.speed = starting_speed;
+      final_points.push_back(pair);  // Store the points in reverse
+      prev_pair = pair;
+      continue;  // continue until loop end
     }
 
-    std::vector<cav_msgs::TrajectoryPlanPoint> StopandWait::compose_trajectory_from_centerline(
-    const std::vector<PointSpeedPair>& points, const cav_msgs::VehicleState& state)
+    double dx = lanelet::geometry::distance2d(prev_pair.point, points[i].point);
+
+    double v_f = sqrt(v_i * v_i + 2 * target_accel * dx);
+
+    PointSpeedPair pair = points[i];
+    pair.speed = std::min(v_f, starting_speed);
+    final_points.push_back(pair);  // Store the points in reverse
+
+    prev_pair = pair;
+  }
+
+  // Now we have a trajectory that decelerates from our end point to somewhere in the maneuver
+  std::reverse(final_points.begin(),
+               final_points.end());  // TODO seems lik final_points does not contain start and end point for some reason
+
+  std::vector<double> speeds;
+  std::vector<lanelet::BasicPoint2d> raw_points;
+  splitPointSpeedPairs(final_points, &raw_points, &speeds);
+
+  std::vector<double> downtracks = carma_wm::geometry::compute_arc_lengths(raw_points);
+
+  double downtrack = 0;
+  bool in_range = false;
+  double stopped_downtrack = 0;
+  lanelet::BasicPoint2d stopped_point;
+  for (size_t i = 0; i < speeds.size(); i++)
+  {  // Apply minimum speed constraint
+    double downtrack = downtracks[i];
+    if (downtrack > downtracks.back() - stop_location_buffer && speeds[i] < config_.crawl_speed + 0.22352)
```

Comment:
Reminder to add the 0.22352 constant to the crawl speed config value or put a comment explaining what it is